sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use thiserror::Error;
+
+/// Error related to the `Emulator` trait. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "Emulator error: {}".  So including the words "error" or
+/// "Emulator" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Timeout during performing command: {0} timeout: {1}, elsped: {2}")]
+    Timeout(String, u64, u64),
+    #[error("Can't restore resource to initial state: {0}")]
+    ResetError(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum EmuValue {
+    Empty,
+    String(String),
+    FilePath(String),
+    StringList(Vec<String>),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+pub enum EmuState {
+    Off,
+    On,
+    Busy,
+    Error,
+}
+
+/// A trait which represents a `Emulator` instance
+pub trait Emulator {
+    /// Acquires state of emulator
+    fn get_state(&self) -> Result<EmuState>;
+
+    /// Waits for the specified amount of time passed in `timeout_ms`
+    /// for the Emulator state to change to the value provided in parameter `state`.
+    fn wait_for_state(&self, timeout_ms: u64, state: EmuState) -> Result<()>;
+
+    /// Start emulator with provided arguments. Parameter `timeout_ms`
+    /// describe how long Emulator will wait for emulation to start.
+    /// The `reset_resource` parameter describes whether the resources are to be restored
+    /// to their initial state before starting emulated DUT.
+    /// If the parameter value is true, the file will be restored to its original content.
+    fn start(
+        &self,
+        timeout_ms: u64,
+        reset_resource: bool,
+        args: &HashMap<String, EmuValue>,
+    ) -> Result<()>;
+
+    /// Stop emulator with provided arguments. Parameter `timeout_ms`
+    /// describe how long Emulator will wait for emulation to stop.
+    fn stop(&self, timeout_ms: u64) -> Result<()>;
```

Comment:
nit: All the `timeout_ms: u64` should instead be `timeout: Duration`, this will make it obvious when looking at a call site `emulator.stop(2000)` what the unit is.